File: app/src/main/java/cz/ackee/sample/model/rest/ApiDescriptionImpl.java
package cz.ackee.sample.model.rest;

import java.util.Arrays;
import java.util.List;
import java.util.Random;

import cz.ackee.rxoauth.ICredentialsModel;
import cz.ackee.sample.model.LoginResponse;
import cz.ackee.sample.model.SampleItem;
import io.reactivex.Observable;
import okhttp3.MediaType;
import okhttp3.ResponseBody;
import okio.BufferedSource;
import retrofit2.HttpException;
import retrofit2.Response;

/**
 * TODO add class description
 * Created by David Bilik[david.bilik@ackee.cz] on {06/08/16}
 **/
public class ApiDescriptionImpl implements ApiDescription {
    public static final String TAG = ApiDescriptionImpl.class.getName();

    @Override
    public Observable<LoginResponse> login(String name, String passwd) {
        return Observable.just(new LoginResponse("David", "123", "456"));
    }

    @Override
    public Observable<List<SampleItem>> getData() {
        final boolean[] first = {true};
        return Observable.just(Arrays.asList(new SampleItem("a"), new SampleItem("b"), new SampleItem("c")))
                .flatMap(list -> {
                    if (first[0] && new Random().nextInt() % 2 == 0) {
                        first[0] = false;
                        //simulate 401 response
                        return Observable.error(new HttpException(Response.error(401, new ResponseBody() {
                            @Override
                            public MediaType contentType() {
                                return null;
                            }

                            @Override
                            public long contentLength() {
                                return 0;
                            }

                            @Override
                            public BufferedSource source() {
                                return null;
                            }
                        })));
                    } else {
                        return Observable.just(list);
                    }
                });
    }

    @Override
    public Observable<ICredentialsModel> refreshAccessToken(String refreshToken) {
        final boolean[] first = {true};
        return Observable.just(new LoginResponse("David", "123", "456"))
                .flatMap(data -> {
                    if (first[0] && new Random().nextInt() % 2 == 0) {
                        first[0] = false;
                        //simulate 400 response
                        return Observable.error(new HttpException(Response.error(400, new ResponseBody() {
                            @Override
                            public MediaType contentType() {
                                return null;
                            }

                            @Override
                            public long contentLength() {
                                return 0;
                            }

                            @Override
                            public BufferedSource source() {
                                return null;
                            }
                        })));
                    } else {
                        return Observable.just(data);
                    }
                })
                .map(resp -> resp);
    }
}

File: app/src/main/java/cz/ackee/sample/model/rest/ApiDescriptionImpl.kt
package cz.ackee.sample.model.rest

import cz.ackee.rxoauth.OauthCredentials
import cz.ackee.sample.model.LoginResponse
import cz.ackee.sample.model.SampleItem
import io.reactivex.Completable
import io.reactivex.Single
import okhttp3.MediaType
import okhttp3.ResponseBody
import okio.BufferedSource
import retrofit2.HttpException
import retrofit2.Response
import java.util.Arrays
import java.util.Random

/**
 * Implementation of [ApiDescription]
 */
class ApiDescriptionImpl : ApiDescription {

    override fun login(name: String, passwd: String): Single<LoginResponse> {
        return Single.just(LoginResponse("David", "123", "456"))
    }

    override fun getData(): Single<List<SampleItem>> {
        val first = booleanArrayOf(true)
        return Single.just(Arrays.asList(SampleItem("a"), SampleItem("b"), SampleItem("c")))
                .flatMap { list ->
                    if (first[0] && Random().nextInt() % 2 == 0) {
                        first[0] = false
                        //simulate 401 response
                        Single.error(HttpException(Response.error<List<SampleItem>>(401, object : ResponseBody() {
                            override fun contentType(): MediaType? {
                                return null
                            }

                            override fun contentLength(): Long {
                                return 0
                            }

                            override fun source(): BufferedSource? {
                                return null
                            }
                        })))
                    } else {
                        Single.just(Arrays.asList(SampleItem("a"), SampleItem("b"), SampleItem("c")))
                    }
                }
    }

    override fun refreshAccessToken(refreshToken: String): Single<OauthCredentials> {
        val first = booleanArrayOf(true)
        return Single.just(LoginResponse("David", "123", "456"))
                .flatMap { data ->
                    if (first[0] && Random().nextInt() % 2 == 0) {
                        first[0] = false
                        //simulate 400 response
                        Single.just(LoginResponse("David", "123", "456"))
                                .flatMap {
                                    Single.error<OauthCredentials>(HttpException(Response.error<OauthCredentials>(400, object : ResponseBody() {
                                        override fun contentType(): MediaType? {
                                            return null
                                        }

                                        override fun contentLength(): Long {
                                            return 0
                                        }

                                        override fun source(): BufferedSource? {
                                            return null
                                        }
                                    })))
                                }
                    } else {
                        Single.just(LoginResponse("David", "123", "456"))
                                .map { data }
                    }
                }
    }

    override fun logout(): Completable {
        return Completable.complete()
    }

    override fun something(): Single<String> {
        return Single.just("")
    }
}
